sync/sync.go

```diff
@@ -602,55 +589,233 @@ func (s *Synchronizer) fetchAndStorePending(ctx context.Context) error {
+//
+// Consumer of preLatestChan will receive each preLatest once. It is possible preLatest block will skipped,
+// incase latest for the pre-latest is already published.
+// Incase fetched pre-latest belongs to future block, it will be cached and served later.
+func (s *Synchronizer) fetchPreLatest(
+	ctx context.Context,
+	head *core.Block,
+	seenBlockHashes map[felt.Felt]*core.PreLatest,
+	preLatestChan chan *core.PreLatest,
+) {
+	highestBlockHeader := s.highestBlockHeader.Load()
+	if highestBlockHeader == nil || highestBlockHeader.Number > head.Number {
+		// not at the tip of the chain yet, no need to poll pre_latest
+		return
 	}
 
-	preConfirmedBlock, err := s.dataSource.PreConfirmedBlockByNumber(ctx, highestBlockHeader.Number+1)
-	if err != nil {
-		return err
+	if val, hit := seenBlockHashes[*head.Hash]; hit {
+		// Already seen this pre_latest for this head.
+		val.Block.Number = head.Number + 1
+		delete(seenBlockHashes, *head.Hash)
+		select {
+		case <-ctx.Done():
+			return
+		case preLatestChan <- val:
+			return
+		}
 	}
 
-	return s.StorePreConfirmed(&preConfirmedBlock)
-}
+	preLatestPollTicker := time.NewTicker(s.pendingPollInterval)
+	defer preLatestPollTicker.Stop()
 
-func (s *Synchronizer) StartingBlockNumber() (uint64, error) {
-	if s.startingBlockNumber == nil {
-		return 0, errors.New("not running")
+	for {
+		pending, err := s.dataSource.BlockPending(ctx)
+		if err == nil {
+			preLatest := core.PreLatest(pending)
+			// Seen future hash, new latest is on flight
+			// Cache the pre-latest to use when its predecessor is available
+			if *preLatest.Block.ParentHash != *head.Hash {
+				seenBlockHashes[*preLatest.Block.ParentHash] = &preLatest
+				return
+			}
+
+			preLatest.Block.Number = head.Number + 1
+			select {
+			case <-ctx.Done():
+				return
+			case preLatestChan <- &preLatest:
+				return
+			}
+		}
+
+		s.log.Debugw("Error while trying to poll pre_latest block", "err", err)
+		select {
+		case <-ctx.Done():
+			return
+		case <-preLatestPollTicker.C:
+			continue
+		}
 	}
-	return *s.startingBlockNumber, nil
 }
 
-func (s *Synchronizer) HighestBlockHeader() *core.Header {
-	return s.highestBlockHeader.Load()
-}
+// pollPreConfirmed fetches and stores pre_confirmed blocks.
+func (s *Synchronizer) pollPreConfirmed(ctx context.Context) {
+	if s.preConfirmedPollInterval == time.Duration(0) {
+		s.log.Infow("Pre-confirmed block polling is disabled")
+		return
+	}
 
-func (s *Synchronizer) SubscribeNewHeads() NewHeadSubscription {
-	return NewHeadSubscription{s.newHeads.Subscribe()}
-}
+	newHeadSub := s.newHeads.SubscribeKeepLast()
+	defer newHeadSub.Unsubscribe()
 
-func (s *Synchronizer) SubscribeReorg() ReorgSubscription {
-	return ReorgSubscription{s.reorgFeed.Subscribe()}
+	preLatestChan := make(chan *core.PreLatest)
+	defer close(preLatestChan)
+	var preLatestWg stdsync.WaitGroup
+	preLatestWg.Go(func() { s.pollPreLatest(ctx, preLatestChan) })
+
+	mostRecentPredecessor := int64(-1)
```

Comment:
Do we need it to be -1 for when we are syncing from block 0? Is this something that can happen in a real usage of Juno (independent of network).